text/0000-structural-records.md

```diff
@@ -0,0 +1,1262 @@
+contains all the fields you want to pass is not a particularly ergonomic solution.
+Instead, it is rather heavy weight:
+
+```rust
+struct Color {
+    red: u8,
+    green: u8,
+    blue: u8,
+}
+
+fn caller() {
+    ...
+    let color = Color { red: 255, green: 0, blue: 0 };
+    ...
+    do_stuff_with(color);
+    ...
+}
+
+fn do_stuff_with(color: Color) { // Only time we use `Color`!
+    some_stuff(color.red);
+    ...
+    other_stuff(color.green);
+    ...
+    yet_more_stuff(color.blue);
+}
+```
+
+To remedy the ergonomics problem, you may instead opt for using a positional
+tuple to contain the values you want to pass. However, now you have likely
+created a regression for readers of your code since the fields of tuples are
+accessed with their positional index, which does not carry clear semantic intent:
+
+```rust
+fn caller() {
+    ...
+    let color = (255, 0, 0);
+    ...
+    do_stuff_with(color); // Unclear what each position means...
+    ...
+}
+
+fn do_stuff_with(color: (u8, u8, u8)) { // More ergonomic!
+    some_stuff(color.0); // But less readable... :(
+    ...
+    other_stuff(color.1);
+    ...
+    yet_more_stuff(color.2);
+}
+```
+
+Using structural records, we can have our cake and eat it too:
+
+```rust
+fn caller() {
+    ...
+    let color = { red: 255, green: 0, blue: 0 };
+    ...
+    do_stuff_with(color); // ...but here it is clear.
+    ...
+}
+
+fn do_stuff_with(color: { red: u8, green: u8, blue: u8 }) { // More ergonomic!
+    some_stuff(color.red); // *And* readable! :)
+    ...
+    other_stuff(color.green);
+    ...
+    yet_more_stuff(color.blue);
+}
+```
+
+In the above snippet, the semantic intent of the fields is clear both when
+reading the body of the function, as well as when reading the documentation
+when `do_stuff_with` is exposed as a public API.
+
+[@eternaleye]: https://internals.rust-lang.org/t/pre-rfc-unnamed-struct-types/3872/58
+[@kardeiz]: https://internals.rust-lang.org/t/pre-rfc-unnamed-struct-types/3872/65
+
+Another example of reducing boilerplate was given by [@eternaleye]:
+
+```rust
+struct RectangleClassic {
+    width: u64,
+    height u64,
+    red: u8,
+    green: u8,
+    blue: u8,
+}
+
+struct RectangleTidy {
+    dimensions: {
+        width: u64,
+        height: u64,
+    },
+    color: {
+        red: u8,
+        green: u8,
+        blue: u8,
+    },
+}
+```
+
+In the second type `RectangleTidy`, we keep boilerplate to a minimum and we can
+also treat `rect.color` and `rect.dimensions` as separate objects and move them
+out of `rect : RectangleTidy` as units, which we cannot do with `RectangleClassic`.
+If we wanted to do that, we would have to invent two new types `Dimensions` and
+`Color` and then `#[derive(..)]` the bits and pieces that we need.
+As noted by [@kardeiz], this ability may also be useful for serializing one-off
+structures with `serde`.
+
+## Major: Better rapid prototyping and refactoring
```

Comment:
Typo `Major` -> `Minor` here and above.

Reply:
We can disagree but it's not a typo in the sense that I meant something different or made a grammatical error...

Reply:
<sub>We can disagree whether the joke was successful or not, but it was a joke.</sub>

Reply:
For the purposes of review, let's just call this resolved ;)